bundle/hash.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2020 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package bundle
+
+import (
+	"crypto/md5"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"sort"
+)
+
+// HashingAlgorithm represents a subset of hashing algorithms implemented in Go
+type HashingAlgorithm string
+
+// Supported values for HashingAlgorithm
+const (
+	MD5       HashingAlgorithm = "MD5"
+	SHA1      HashingAlgorithm = "SHA1"
+	SHA224    HashingAlgorithm = "SHA224"
+	SHA256    HashingAlgorithm = "SHA256"
+	SHA384    HashingAlgorithm = "SHA384"
+	SHA512    HashingAlgorithm = "SHA512"
+	SHA512224 HashingAlgorithm = "SHA512_224"
+	SHA512256 HashingAlgorithm = "SHA512_256"
+)
+
+// String returns the string representation of a HashingAlgorithm
+func (alg HashingAlgorithm) String() string {
+	return string(alg)
+}
+
+// SignatureHasher computes a signature digest for a file with (structured or unstructured) data and policy
+type SignatureHasher interface {
+	HashFile(v interface{}) ([]byte, error)
+}
+
+type hasher struct {
+	h func() hash.Hash // hash function factory
+}
+
+// NewSignatureHasher returns a signature hasher suitable for a particular hashing algorithm
+func NewSignatureHasher(alg HashingAlgorithm) (SignatureHasher, error) {
+	h := &hasher{}
+
+	switch alg {
+	case MD5:
+		h.h = md5.New
+	case SHA1:
+		h.h = sha1.New
+	case SHA224:
+		h.h = sha256.New224
+	case SHA256:
+		h.h = sha256.New
+	case SHA384:
+		h.h = sha512.New384
+	case SHA512:
+		h.h = sha512.New
+	case SHA512224:
+		h.h = sha512.New512_224
+	case SHA512256:
+		h.h = sha512.New512_256
+	default:
+		return nil, fmt.Errorf("unsupported hashing algorithm: %s", alg)
+	}
+
+	return h, nil
+}
+
+// HashFile hashes the file content, JSON or binary, both in golang native format.
+func (h *hasher) HashFile(v interface{}) ([]byte, error) {
+	return h.file(v)
+}
+
+// file hashes the file content, JSON or binary, both in golang native format.
+
+// Computation for unstructured documents is a hash of string "binary",
+// hashed together with a content digest of the document.
+
+// Computation for the types of structured JSON document is as follows:
+//
+// null: Hash null
+//
+// string: Hash "string" (to signify type) and then hash the string value.
+//
+// number: Hash "number" (to signify type) and then hash the number value (as a string).
+//
+// boolean: Hash "boolean" (to signify type) and then hash the true or false.
+//
+// object: Hash {, then each key (in alphabetical order) and digest of the value, and finally }.
+//
+// array: Hash [, then digest of the value, and finally ].
+func (h *hasher) file(v interface{}) ([]byte, error) {
```

Comment:
Per the offline discussion, we can align this algorithm to match the hashing of raw (ordered) JSON file by making the actual digested content to match the bytes matched in that case. In other words, need to remove a) type strings from hashed, b) add commas to containers, c) consider the primitive types' hashing details.

Reply:
Updated the algorithm.